p2p/transport/tcp/tcp.go

```diff
@@ -113,6 +117,13 @@ func WithMetrics() Option {
 	}
 }
 
+func WithSharedTCP(mgr *tcpreuse.ConnMgr) Option {
```

Comment:
There's a few API questions on how we want to expose this functionality. For example:
1. Is it an option passed into the transport constructors for TCP + WebSockets, or is it one of the parameters controlled by fx and if nil is passed we don't allow shared listeners (this would lend itself better to the SharedTCP option being enabled at the host constructor layer)?
2. Should shared listeners be enabled by default in the host constructor (whether now or in the future)?